src/Features/CSharp/Portable/UsePatternMatching/CSharpAsAndNullCheckDiagnosticAnalyzer.cs

```diff
@@ -195,71 +240,88 @@ private void SyntaxNodeAction(SyntaxNodeAnalysisContext syntaxContext)
+                    if (isAssigned)
+                    {
+                        return true;
+                    }
+
+                    if (isAccessedBeforeAssignment)
+                    {
+                        // Access before assignment here is only valid if we have a negative
+                        // pattern-matching in an if-statement with an unreachable endpoint.
+                        // For example:
+                        //
+                        //      if (!(o is string x)) {
+                        //        return;
+                        //      }
+                        //
+                        //      // The 'return' statement above ensures x is definitely assigned here
+                        //      Console.WriteLine(x);
+                        //
+                        return isNegativeNullCheck && 
+                           ifStatement != null && 
+                           !semanticModel.AnalyzeControlFlow(ifStatement.Statement).EndPointIsReachable;
+                    }
                 }
             }
 
-            return false;
+            return true;
         }
 
         private static void CheckDefiniteAssignment(
-            SemanticModel semanticModel, ISymbol localVariable, SyntaxNode node,
-            out bool isAssigned, out bool isAccessedBeforeAssignment,
-            CancellationToken cancellationToken)
+            SemanticModel semanticModel, ISymbol localVariable, 
+            StatementSyntax firstStatement, StatementSyntax lastStatement,
+            out bool isAssigned, out bool isAccessedBeforeAssignment)
         {
-            if (node != null)
+            if (firstStatement is null)
             {
-                foreach (var descendantNode in node.DescendantNodes())
-                {
-                    if(!descendantNode.IsKind(SyntaxKind.IdentifierName, out IdentifierNameSyntax id))
-                    {
-                        continue;
-                    }
-
-                    var symbol = semanticModel.GetSymbolInfo(id, cancellationToken).GetAnySymbol();
-                    if (localVariable.Equals(symbol))
-                    {
-                        isAssigned = id.IsOnlyWrittenTo();
-                        isAccessedBeforeAssignment = !isAssigned;
-                        return;
-                    }
-                }
+                isAssigned = false;
+                isAccessedBeforeAssignment = false;
+                return;
             }
 
-            isAssigned = false;
-            isAccessedBeforeAssignment = false;
+            var dataFlow = lastStatement is null
+                ? semanticModel.AnalyzeDataFlow(firstStatement)
+                : semanticModel.AnalyzeDataFlow(firstStatement, lastStatement);
+            isAssigned = dataFlow.AlwaysAssigned.Contains(localVariable);
+            isAccessedBeforeAssignment = !isAssigned && dataFlow.ReadInside.Contains(localVariable);
```

Comment:
There is a bug here, should be missing but it's not:
```cs
    void M()
    {
        [|var|] x = o as string;
        if (x != null)
        {
        }
        Console.WriteLine(x);
        x = "";
    }
}
```
x is always but there is a read before assignment - probably need to revert code to iterate over statements one by one, @CyrusNajmabadi, can you confirm if that was the original purpose of doing so instead of region analysis?

Reply:
> There is a bug here, should be missing but it's not:

It's missing for me.  I'm not able to see this locally.

Reply:
I feel there should be a combination of DataFlowsIn/Out that helps.

Reply:
Note: iterating over statements is not safe because assignments could be anywhere in an expression,
```cs
Console.WriteLine( Read(x) ??  (x = "tricky"));
```

Reply:
@CyrusNajmabadi I meant in this branch. The previous revision check for nodes statement so doesn't have this bug.

However, with either approaches, this one is not missing.
```cs
    void Use(string x) {}
    void M()
    {
        var x = o as string;
        if (x != null)
        {
        }
        Console.WriteLine(x = Use(x));
    }
```
From data flow analysis results, we can't see if the read is happened before write..

Reply:
>  @CyrusNajmabadi, can you confirm if that was the original purpose of doing so instead of region analysis?

Yes, i believe that was the intent. 

> Note: iterating over statements is not safe because assignments could be anywhere in an expression,

I thought we checked assignment on a per-reference basis.  So we would specifically check the Use(x) and see the problem.  Right?

Reply:
I think we return early in the previous code. we see the write first and bail. I'll try a mixed approach, if we are AlwaysAssgined, check if there is a read, if yes, need to walk through tree to see it doesn't come first. early return here could be safe.

Reply:
Oh, i think i see what you're saying.  The logic here is buggy:

https://github.com/dotnet/roslyn/blob/f5afdc19c154e9a93561fa66ec013c2cd7730485/src/Features/CSharp/Portable/UsePatternMatching/CSharpAsAndNullCheckDiagnosticAnalyzer.cs#L230-L233

Specifically, because it sees the ```x =```, it assumes that 'x' is always assigned, and thus the change is safe.  

1. That sucks.  But it's also a very corner case.
2. We could consider fixing it.  Namely that if the reference is the left of an assign, that we have to ensure that the right side is fine as well.  Annoying, but doable.

Reply:
Note: using AlwaysAssigned is necessary to handle branches e.g. https://github.com/dotnet/roslyn/issues/21097

Reply:
I see.  The logic here has changed a bit, and i'll admit i don't really fully understand the algorithm as it stands now, so i'm not sure the exact right fix.  But i do think this can be solved.

Also, an efficiency point:

```c#
                    var symbol = semanticModel.GetSymbolInfo(id, cancellationToken).GetAnySymbol();
                    if (localVariable.Equals(symbol))
                    {
```

We should check if the id.ValueText matches the local variable name before bothering to do the binding.  That will avoid a lot of expensive semantic checks.